### PR TITLE
add alignment setting per shape system

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

- Added alignment configuration option for shape systems. That allows creating bottom-left aligned sprites, which will behave much more naturally inside auto-layouts.
- Added support for alignment in manual layouts. When alignment is set, the manual layout will position the child node at the respective border of its bounding box. The size of aligned node will not be affected. The difference from auto-layout alignment is that each node is aligned individually, and it is not affected by its siblings. This allows for constructing more complicated responsive layouts that don't necessarily follow the grid, without creating wrapper auto-layout elements for each child.

layout-anim-and-alignment

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: lib/rust/ensogl/core/src/display/object/instance.rs [hunk 1]
```diff
@@ -3108,48 +3115,103 @@ impl Model {
...
+                    let aligned_position = remaining_size * alignment.normalized();
+                    child.set_position_dim(x, aligned_position);
+                    let child_content_origin = child.content_origin().get_dim(x);
+                    let child_min_x = aligned_position + child_content_origin;
+                    let child_max_x = child_min_x + child_size;
+                    min_x = min_x.min(child_min_x);
+                    max_x = max_x.max(child_max_x);
+                }
+            }
+
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/object/instance.rs]: The comment in line 3162 applies to lines 3180-3181 as well, right? If so, I'd delete this nl.

File: lib/rust/ensogl/core/src/display/shape/primitive/system.rs [hunk 2]
```diff
@@ -446,10 +452,12 @@ impl ShapeSystemModel {
...
-                // We need to recompute the vertex position with the padding.
-                input_local = vec3((input_uv - input_alignment) * input_size, 0.0);
-                gl_Position = model_view_projection * vec4(input_local,1.0);
-                input_local.z = gl_Position.z;
+                // Compute the vertex position with shape padding, apply alignment to the vertex
+                // position.
+                vec2 position = vec2((input_uv - input_alignment) * input_size);
+                gl_Position = model_view_projection * vec4(position,0.0,1.0);
+                // Compute SDF input position with 0.0 being in the shape's center.
+                input_local = vec3((input_uv - vec2(0.5)) * input_size, gl_Position.z);
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/shape/primitive/system.rs]: Please double check if the growing border (`ctrl + alt + 5`) is correct after these changes for every alignment. It is a very tricky part and needs to be double checked. Post screenshots of it please if you can.

**Review by wdanilo — APPROVED** [review]
Beautiful! One important question - you've written that with a left-alignment auto-layout behaves "better"? Auto-layout should not behave differently no matter the children alignment. If it does, there is a bug.

**Comment by wdanilo** [comment]
Also, could you change the default alignment of the super-box shape to bottom-left and check if everything works correctly, please? We should slowly move to all things having bottom-left alignment. The centered alignment was probably a design mistake in early EnsoGL days.

File: lib/rust/ensogl/core/src/display/shape/primitive/system.rs (hunk 2 above)
Comment (Frizi) [reply lib/rust/ensogl/core/src/display/shape/primitive/system.rs]: Growing border:
image


AA regions:
image

This scene contains shapes that are both center aligned (top group), and bottom-left aligned (bottom group). Their padding is now displayed correctly. The layout differences are due to `content_origin` being set to negative values in center-padded shapes, which effectively leads to their siblings or children being offset. It's a separate topic that will be solved in different PR.

It was indeed incorrect to remove the initial transform before `aa_side_padding`. I have a feeling that this additional matrix multiplication is a little bit wasteful. Also close zoom levels, it produces overly big padding (way bigger than 1px). But I don't want to spend time tweaking that now. It only adds to the fill rate a little bit, without having any other noticeable impact.

image


For future reference, there is a nice GPU gems chapter about emulating conservative rasterization for triangles, which might be useful at some point: https://developer.nvidia.com/gpugems/gpugems2/part-v-image-oriented-computing/chapter-42-conservative-rasterization

**Comment by Frizi** [comment]
I'm not sure what are you referring to by "super-box". I'm sure though that globally changing the alignment of existing elements would impact their sprite positions. All of them would need to be repositioned to compensate. Also if we are going to actually change default alignment everywhere, we might as well consider top-left alignment as default, as it is done in DOM. That way the default UI flow would match reading order.

**Comment by wdanilo** [comment]
Regarding `Also close zoom levels, it produces overly big padding (way bigger than 1px).` - I understand that it did work correctly before this PR, right? If so, I'd spend the time fixing it, as it will be really hard to get fixed in some time, when we will forget about the context and how it really works under the hood. This is a regression and I think it prevents this PR from merging. Ofc, if you'd have problem with debugging it, I'd be happy to help however I can.

Regarding "superbox" I mean the new rectangle shape that was recently implemented in `ensogl_core::display::shape::compound::rectangle`

**Comment by Frizi** [comment]
> Regarding Also close zoom levels, it produces overly big padding (way bigger than 1px). - I understand that it did work correctly before this PR, right?

No, this didn't change. It's just an observation I've made while working on this. It behaves identically on develop.

> Regarding "superbox" I mean the new rectangle shape that was recently implemented in ensogl_core::display::shape::compound::rectangle

Okay, I see it's not used yet anywhere, so it makes sense to change the alignment.

File: lib/rust/ensogl/core/src/display/shape/primitive/system.rs [hunk 3]
```diff
@@ -444,12 +455,26 @@ impl ShapeSystemModel {
...
                 vec2 uv_offset = padding / input_size;
-                input_uv = input_uv * uv_scale - uv_offset;
 
-                // We need to recompute the vertex position with the padding.
-                input_local = vec3((input_uv - input_alignment) * input_size, 0.0);
-                gl_Position = model_view_projection * vec4(input_local,1.0);
-                input_local.z = gl_Position.z;
+                // Note: Due to a shader optimizer issue, the `input_uv` modification is
+                // intentionally split into two separate expressions. When this is written in single
+                // line as follows:
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/shape/primitive/system.rs]: no info what issue. Please provide details + link to the ticket

File: lib/rust/ensogl/core/src/display/shape/compound/rectangle.rs [hunk 4]
```diff
@@ -34,8 +36,8 @@ mod shape {
             // canvas area. Thus, we need to recompute the new canvas size for the scaled shape.
             let canvas_clip_height_diff = &canvas_height * (clip.y() * 2.0);
             let canvas_clip_width_diff = &canvas_width * (clip.x() * 2.0);
-            let canvas_height = canvas_height + &canvas_clip_height_diff;
-            let canvas_width = canvas_width + &canvas_clip_width_diff;
+            let canvas_height = canvas_height + &canvas_clip_height_diff.abs();
+            let canvas_width = canvas_width + &canvas_clip_width_diff.abs();
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/shape/compound/rectangle.rs:40]: I was trying to keep the shader as optimized as possible. This implementation gives the same possibilities with slightly slower shader, thus I prefer the previous one.

File: lib/rust/ensogl/core/src/display/shape/primitive/system.rs (hunk 3 above)
Comment (Frizi) [reply lib/rust/ensogl/core/src/display/shape/primitive/system.rs]: added issue link: https://github.com/KhronosGroup/SPIRV-Cross/issues/2129

**Pull request merged**